contracts/base/strategies/beethoven/BeethovenBase.sol

```diff
@@ -0,0 +1,220 @@
+    for (uint256 i = 0; i < poolTokens.length; i++) {
+      token[i] = address(poolTokens[i]);
+    }
+    return token;
+  }
+
+  // ************ GOVERNANCE ACTIONS **************************
+
+  /// @notice Claim rewards from external project and send them to FeeRewardForwarder
+  function doHardWork() external onlyNotPausedInvesting override restricted {
+    investAllUnderlying();
+    IBeethovenxChef(pool).harvest(poolId, address(this));
+    liquidateReward();
+  }
+
+  // ************ INTERNAL LOGIC IMPLEMENTATION **************************
+
+  /// @dev Deposit underlying to MasterChef pool
+  /// @param amount Deposit amount
+  function depositToPool(uint256 amount) internal override {
+    IERC20(_underlyingToken).safeApprove(pool, 0);
+    IERC20(_underlyingToken).safeApprove(pool, amount);
+    IBeethovenxChef(pool).deposit(poolId, amount, address(this));
+  }
+
+  /// @dev Withdraw underlying from MasterChef pool
+  /// @param amount Deposit amount
+  function withdrawAndClaimFromPool(uint256 amount) internal override {
+     IBeethovenxChef(pool).withdrawAndHarvest(poolId, amount, address(this));
+  }
+
+  /// @dev Exit from external project without caring about rewards
+  ///      For emergency cases only!
+  function emergencyWithdrawFromPool() internal override {
+    IBeethovenxChef(pool).emergencyWithdraw(poolId, address(this));
+  }
+
+  /// @dev Do something useful with farmed rewards
+  function liquidateReward() internal override {
+    autoCompoundBalancer();
+    liquidateRewardSilently();
+  }
+
+  /// @dev Concert pool tokens from IERC20[] to IAsset[]
+  function setupPullTokens() internal {
+    (IERC20[] memory tokens,,) = beethovenVault.getPoolTokens(beethovenPoolId);
+    IAsset[] memory tokenAssets = new IAsset[](tokens.length);
+    for (uint256 i = 0; i < tokens.length; i++) {
+      tokenAssets[i] = IAsset(address(tokens[i]));
+    }
+    poolTokens = tokenAssets;
+  }
+
+  /// @dev Liquidate rewards, buy assets and add to beethoven pool
+  function autoCompoundBalancer() internal {
+    for (uint256 i = 0; i < _rewardTokens.length; i++) {
+      uint256 amount = rewardBalance(i);
+      if (amount != 0) {
+        uint toCompound = amount * (_BUY_BACK_DENOMINATOR - _buyBackRatio) / _BUY_BACK_DENOMINATOR;
+        address rt = _rewardTokens[i];
+        rewardToUnderlying(rt, toCompound);
+        depositToPool(underlyingBalance());
+      }
+    }
+  }
+
+  /// @dev swap reward token to underlying using Beethoven pool
+  function rewardToUnderlying(address rewardToken, uint toCompound) internal{
+    if (toCompound == 0) {
+      return;
+    }
+    if (rewardToken != depositToken) {
+      balancerSwap(rewardToDepositPoolId, rewardToken, depositToken, toCompound);
+    }
+    uint depositTokenBalance = IERC20(depositToken).balanceOf(address (this));
+    IERC20(depositToken).safeApprove(address(beethovenVault), 0);
+    IERC20(depositToken).safeApprove(address(beethovenVault), depositTokenBalance);
```

Comment:
better to move approves inside `balancerJoin`

Reply:
done